### PR TITLE
Wait for PG Deployment Deletion on Restore

### PR DESCRIPTION
**Pull request by andrewlecuyer**
Since the primary PVC for the cluster is now retained during an in-place PostgreSQL cluster restore in support of pgBackRest delta restores, when preparing a cluster for a restore we can no longer rely on the deletion of all PVC's as an indicator that the `config` and `leader` ConfigMaps created by Patroni can be removed.  Therefore, the Operator now specifically waits for all Deployments to be successfully removed prior to deleting these resources.

**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [ ] Have you updated or added documentation for the change, as applicable?
 - [x] Have you tested your changes on all related environments with successful results, as applicable?

**Type of Changes:**

 
 - [x] Bug fix (non-breaking change which fixes an issue)
 - [ ] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change)

**What is the current behavior? (link to any open issues here)**

The deletion of all primary and replica PVCs is used to indicate that it is safe to remove the `config` and `leader` ConfigMaps when performing an in-place cluster restore.

Issue: [ch9926]

**What is the new behavior (if this is a feature change)?**

The Operator now specifically waits for all Deployments to be successfully removed prior to deleting the `config` and `leader` ConfigMaps when performing an in-place cluster restore.

**Other information**:

N/A

### CODE REVIEW

File: internal/operator/backrest/restore.go [hunk 1]
```diff
@@ -184,7 +184,25 @@ func PrepareClusterForRestore(clientset kubeapi.Interface, cluster *crv1.Pgclust
...
+		clusterName)
+
+	// Wait for all primary and replica deployments to be removed.  If unable to verify that all
+	// deployments have been removed, then the restore cannot proceed and the function returns.
+	if err := wait.Poll(time.Second/2, time.Minute*3, func() (bool, error) {
+		notFound := true
+		for _, deployment := range pgInstances.Items {
+			if _, err := clientset.AppsV1().Deployments(namespace).
+				Get(ctx, deployment.GetName(), metav1.GetOptions{}); err == nil {
+				notFound = false
```
Comment (jkatz) [inline internal/operator/backrest/restore.go:197]: We could just `return false, nil` here?

File: internal/operator/backrest/restore.go [hunk 2]
```diff
@@ -184,7 +184,25 @@ func PrepareClusterForRestore(clientset kubeapi.Interface, cluster *crv1.Pgclust
...
+			}
+		}
+		return notFound, nil
```
Comment (jkatz) [inline internal/operator/backrest/restore.go:200]:
```suggestion
		for _, deployment := range pgInstances.Items {
			if _, err := clientset.AppsV1().Deployments(namespace).
				Get(ctx, deployment.GetName(), metav1.GetOptions{}); err == nil {
				return false, nil
			}
		}
		return true, nil
```

File: internal/operator/backrest/restore.go [hunk 3]
```diff
@@ -184,7 +184,25 @@ func PrepareClusterForRestore(clientset kubeapi.Interface, cluster *crv1.Pgclust
...
+	log.Debugf("restore workflow: deleted primary and replica deployments for cluster %s",
+		clusterName)
+
+	// Wait for all primary and replica deployments to be removed.  If unable to verify that all
+	// deployments have been removed, then the restore cannot proceed and the function returns.
+	if err := wait.Poll(time.Second/2, time.Minute*3, func() (bool, error) {
+		notFound := true
+		for _, deployment := range pgInstances.Items {
+			if _, err := clientset.AppsV1().Deployments(namespace).
+				Get(ctx, deployment.GetName(), metav1.GetOptions{}); err == nil {
```
Comment (cbandy) [inline internal/operator/backrest/restore.go:196]: Might want to check for NotFound here. Other errors can be retried.

Comment (andrewlecuyer) [reply internal/operator/backrest/restore.go]: Now specifically checking for `IsNotFound`.

File: internal/operator/backrest/restore.go (hunk 1 above)
Comment (andrewlecuyer) [reply internal/operator/backrest/restore.go]: Recommendation included in latest push.

File: internal/operator/backrest/restore.go (hunk 2 above)
Comment (andrewlecuyer) [reply internal/operator/backrest/restore.go]: Recommendation included in latest push.

**Review by jkatz — APPROVED** [review]

**Pull request merged**